lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -1054,31 +1054,82 @@ def _stop_opensearch(self, *, restart=False) -> None:
         """Stop OpenSearch if possible."""
         self.status.set(WaitingStatus(ServiceIsStopping))
 
-        if self.opensearch.is_node_up():
-            try:
-                nodes = self._get_nodes(True)
-                # do not add exclusions if it's the last unit to stop
-                # otherwise cluster manager election will be blocked when starting up again
-                # and re-using storage
-                if len(nodes) > 1:
-                    # TODO: we should probably NOT have any exclusion on restart
-                    # https://chat.canonical.com/canonical/pl/bgndmrfxr7fbpgmwpdk3hin93c
-                    # 1. Add current node to the voting + alloc exclusions
-                    self.opensearch_exclusions.add_current()
-            except OpenSearchHttpError:
-                logger.debug("Failed to get online nodes, voting and alloc exclusions not added")
+        if not self.opensearch.is_started():
+            self.status.set(WaitingStatus(ServiceStopped))
+            # Nothing to do, the service is not yet up
+            return
 
-        # TODO: should block until all shards move addressed in PR DPE-2234
+        # 1. stop the service
+        try:
+            self.opensearch.stop()
+            self.peers_data.delete(Scope.UNIT, "started")
+            self.status.set(WaitingStatus(ServiceStopped))
+        except OpenSearchStopError as e:
+            logger.exception(e)
+            # Now, we need to decide if the service is still functional or not
+            if not self.opensearch.is_node_up():
+                # We are still running, there was an error
+                raise
 
-        # 2. stop the service
-        self.opensearch.stop()
-        self.peers_data.delete(Scope.UNIT, "started")
-        self.status.set(WaitingStatus(ServiceStopped))
+        # 2. Check if we have enough voting units
+        self._settle_voting_exclusions(exclude_this_unit=True)
```

Comment:
Is it a good idea to work with the API after the service has been stopped?

Reply:
This should target every unit

Reply:
Moved it to be executed before stopping the unit.